utils.py

```diff
@@ -111,30 +116,78 @@ def make_confusion_matrix(y_true, y_pred, classes=None, figsize=(10, 10), text_s
                 size=text_size,
             )
     plt.show()
+    # path = str()
+    # for name in classes:
+    #     path += name.replace(' ', '-').lower()
+    # path += '.png'
+    # plt.savefig(f"plots/{path}")
 
 
 def test_model(model: nn.Module, test_data: DataLoader):
     """
-    Test the performance of the trained ComposerClassifier model with the provided test data loader
-    .by plotting a confusion matrix.
+    Test the performance of the trained ComposerClassifier model with the provided test data loader.
```

Comment:
```suggestion
    Creates a shuffled copy of the input dataset in as a dict, with additional 'pred' column for model predictions.
```

I think this would be a more accurate description of the function :)

This goes beyond this function, but here's an approach that I think would be better:

```python
test_loader = DataLoader(test_dataset, shuffle=False)
predictions = []
for batch in test_loader:
  out = model(batch["data"])
  batch_prediction = out.argmax(1)
  predictions.append(batch_prediction)

predictions = torch.hstack(predictions)
target = [s["label"] for s in test_dataset]
results = predictions == torch.tensor(target)
```

Here `results` is a boolean array that can be used as indices to the dataset, e.g.:

```python
# Review bad predictions
idxs = np.where(~results)[0]
samples = [test_dataset[idx] for idx in idxs]
...

# Review per-piece performance
df = pd.DataFrame([s["title"] + "_" + s["composer"] for s in test_dataset], columns=["piece"])
df["result"] = results
groups = df.groupby("piece")

for piece, gdf in groups:
  print(piece, gdf.result.mean())
```

That way you can see the performance of your model on a piece-by-piece basis and find outliers:

```
Piece: Sonata No. 3 in C Major, Op. 2 No. 3, First Movement_Ludwig van Beethoven
Accuracy: 0.94
---
Piece: Transcendental Etude No. 3 "Paysage"_Franz Liszt
Accuracy: 0.34
```

Reply:
okay, will do